plugins/outputs/splunk/splunk.go

```diff
@@ -0,0 +1,277 @@
+		fields := newTags
+		for fieldName, value := range m.Fields() {
+			fields["metric_name"] = processFieldName(m.Name(), fieldName, s)
+
+			metricValue, buildError := processFieldValue(value, fieldName, s)
+			if buildError != nil {
+				log.Printf("D! [Splunk] %s\n", buildError.Error())
+				continue
+			}
+			fields["_value"] = metricValue
+
+			// -----------------------------------------------------------------------------------------------------------------
+			//  Create a []byte array to send via an HTTP POST
+			// -----------------------------------------------------------------------------------------------------------------
+			var payload []byte
+			var err error
+
+			splunkMetric.Time = m.Time().UnixNano() / 1000000000 // convert metric nanoseconds to unix time
+			splunkMetric.Event = "metric"
+			splunkMetric.Fields = fields
+			payload, err = json.Marshal(splunkMetric)
+
+			if err != nil {
+				return fmt.Errorf("unable to marshal data, %s\n", err.Error())
+			}
+			log.Printf("D! Output [Splunk] %s\n", payload)
+
+			// -----------------------------------------------------------------------------------------------------------------
+			//  Send the data to Splunk
+			// -----------------------------------------------------------------------------------------------------------------
+			req, err := http.NewRequest("POST", s.SplunkUrl, bytes.NewBuffer(payload))
+			if err != nil {
+				return fmt.Errorf("unable to create http.Request \n    URL:%s\n\n", s.SplunkUrl)
+			}
+
+			req.Header.Add("Content-Type", "application/json")
+			req.Header.Add("Authorization", "Splunk "+s.AuthString)
+
+			//// Check for existence of s && req to prevent race ('panic: runtime error: invalid memory address or nil pointer dereference')
+			if s != nil && req != nil {
+				resp, err := s.client.Do(req)
+				if err != nil {
+					return fmt.Errorf("error POST-ing metrics to Splunk[%s]  Sending Data:%s\n", err, payload)
+				}
+
+				defer resp.Body.Close()
+
+				if resp.StatusCode < 200 || resp.StatusCode > 209 {
+					return fmt.Errorf("received bad status code posting to %s, %d\n\n%s\n", s.SplunkUrl, resp.StatusCode, payload)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// -----------------------------------------------------------------------------------------------------------------------
+//  Check for metric name ending with .value and remove it, sanitize characters and add separators
+// -----------------------------------------------------------------------------------------------------------------------
+func processFieldName(measureName string, fieldName string, s *Splunk) string {
+	var name string
+	if !s.SimpleFields && fieldName == "value" {
+		name = fmt.Sprintf("%s%s", s.Prefix, measureName)
+	} else {
+		name = fmt.Sprintf("%s%s%s%s", s.Prefix, measureName, s.MetricSeparator, fieldName)
+	}
+
+	if s.UseRegex {
+		name = sanitizedRegex.ReplaceAllLiteralString(name, "-")
+	} else {
+		name = sanitizedChars.Replace(name)
+	}
```

Comment:
I don't think we should be forcing the user to have scrubbed field names. Defaulting to enabled might be fine, but it should be possible to turn it off entirely.
Splunk doesn't require that the name be within a certain character set. Having some characters might make it harder to work with, but it's not impossible. And people might not want their names being messed with.

Reply:
modified the default to NOT change any names and added a new setting to let users choose if they want to replace special characters.